shiny/ui/_markdown_stream.py

```diff
@@ -0,0 +1,276 @@
+                    result += c
+                    await self._send_content_message(c, "append")
+
+            return result
+
+        _task()
+
+        # Since the task runs in the background (outside/beyond the current context,
+        # if any), we need to manually raise any exceptions that occur
+        @reactive.effect
+        async def _handle_error():
+            e = _task.error()
+            if e:
+                await self._raise_exception(e)
+            _handle_error.destroy()  # type: ignore
+
+        return _task
+
+    async def clear(self):
+        """
+        Clear the content of the markdown stream.
+        """
+        return await self.stream([], clear=True)
+
+    @asynccontextmanager
+    async def _streaming_dot(self):
+        await self._send_stream_message(True)
+        try:
+            yield
+        finally:
+            await self._send_stream_message(False)
+
+    async def _send_content_message(
+        self,
+        content: str,
+        operation: Literal["append", "replace"],
+    ):
+        msg: ContentMessage = {
+            "id": self.id,
+            "content": content,
+            "operation": operation,
+        }
+        await self._send_custom_message(msg)
+
+    async def _send_stream_message(self, is_streaming: bool):
+        msg: isStreamingMessage = {
+            "id": self.id,
+            "isStreaming": is_streaming,
+        }
+        await self._send_custom_message(msg)
+
+    async def _send_custom_message(
+        self, msg: Union[ContentMessage, isStreamingMessage]
+    ):
+        if self._session.is_stub_session():
+            return
+        await self._session.send_custom_message("shinyMarkdownStreamMessage", {**msg})
+
+    async def _raise_exception(self, e: BaseException):
+        if self.on_error == "unhandled":
+            raise e
+        else:
+            sanitize = self.on_error == "sanitize"
+            msg = f"Error in MarkdownStream('{self.id}'): {str(e)}"
+            raise NotifyException(msg, sanitize=sanitize) from e
+
+
+@add_example()
+def markdown_stream_ui(
+    id: str,
+    *,
+    content: str = "",
+    content_type: StreamingContentType = "markdown",
+    auto_scroll: bool = True,
+    width: CssUnit = "100%",
+    height: CssUnit = "auto",
+) -> Tag:
+    """
+    Create a UI element for a markdown stream.
+
+    This method is only relevant for Shiny Core. In Shiny Express, use
+    :meth:`~shiny.ui.MarkdownStream.ui` to get the UI element for the markdown stream
+
+    Parameters
+    ----------
+    id
+        A unique identifier for this markdown stream.
+    content
+        Some content to display before any streaming occurs.
+    content_type
+        The content type. Default is "markdown" (specifically, CommonMark).
+        Other supported options are:
+        - `"html"`: for rendering HTML content.
+        - `"text"`: for plain text.
+        - `"semi-markdown"`: for rendering markdown, but with HTML tags escaped.
```

Comment:
Is it possible to change this choice after you've created the UI element? It's worth considering.

Or maybe it's that this feels like it's in the wrong place, since the message is the thing that's in html, text or markdown. I can see this being less relevant for the chat use case but as a standalone component it seems feasible to want to switch between html/text/markdown.

I think we should leave `content_type` here, to pair with `content`, but we should also have `content_type` as an argument in `stream()`. That argument could default to `None` – i.e. don't change the content type – but if provided we'd change the content type with the first message. (Just brainstorming out loud...)

Reply:
> but we should also have content_type as an argument in stream()

I agree, but unfortunately it wouldn't be support that combined with `.stream(clear=False)` since you could end up with a mix of content types. Thats said, I guess we could just not allow `content_type` to be specified when `clear=False`?

I guess another way we could go to get around this is to have `.stream()` always clear/replace the content, and then a `.append_stream()` that wouldn't have `content_type`?